platform/security/idp/security-idp-plugin/security-idp-plugin-subjectconfirmation/src/main/java/org/codice/ddf/security/idp/plugin/subjectconfirmation/SubjectConfirmationPresignPlugin.java

```diff
@@ -0,0 +1,140 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.security.idp.plugin.subjectconfirmation;
+
+import static java.util.Objects.nonNull;
+
+import com.google.common.collect.Maps;
+import ddf.security.samlp.MetadataConfigurationParser;
+import ddf.security.samlp.SamlProtocol.Binding;
+import ddf.security.samlp.impl.EntityInformation;
+import ddf.security.samlp.impl.EntityInformation.ServiceInfo;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import org.codice.ddf.security.idp.plugin.SamlPresignPlugin;
+import org.joda.time.DateTime;
+import org.opensaml.saml.saml2.core.Assertion;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+import org.opensaml.saml.saml2.core.Response;
+import org.opensaml.saml.saml2.core.SubjectConfirmation;
+import org.opensaml.saml.saml2.core.SubjectConfirmationData;
+import org.opensaml.saml.saml2.core.impl.SubjectConfirmationDataBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SubjectConfirmationPresignPlugin implements SamlPresignPlugin {
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(SubjectConfirmationPresignPlugin.class);
+
+  @Override
+  public void processPresign(
+      Response response,
+      AuthnRequest authnRequest,
+      List<String> spMetadata,
+      Set<Binding> supportedBindings) {
+
+    String inResponseTo = response.getInResponseTo();
+
+    final DateTime notOnOrAfter = DateTime.now().plusMinutes(30);
+
+    String acsUrl = getAssertionConsumerServiceURL(authnRequest, spMetadata, supportedBindings);
+
+    for (Assertion assertion : response.getAssertions()) {
+      assertion
+          .getSubject()
+          .getSubjectConfirmations()
+          .forEach(sc -> setConfirmationData(sc, notOnOrAfter, acsUrl, inResponseTo));
+    }
```

Comment:
❓  Why not stream the assertions? 
```
response.getAssertions()
      .stream()
      .map(asrt::getSubject)
      .map(sub::getSubjectConfirmations)
      .flatMap(confirms::stream)
      .forEach(...);
```
I'm actually on the fence if that truly improves readability. It may not. Removes one level of nesting though. 

Another thought I had was throwing all your local variables onto an inner class and leveraging that to optimize your use of method references.

Reply:
I tend to avoid nested streams as they don't read great.

Reply:
I don't believe @Lambeaux's solution would create nested streams since he his flatmapping the secondary collection all within the same stream.

Reply:
I'm OK with either approach. Just throwing out alternatives.